tests/kernel/timer/timer_monotonic/src/main.c

```diff
@@ -8,21 +8,139 @@
 #include <tc_util.h>
 #include <ztest.h>
 
-int test_frequency(void)
+static struct k_timer test_timer;
+static struct k_timer test_periodtimer;
+
+u32_t start_time, stop_time, cycles_spent, nanoseconds_spent,
+	   time_spent[10], cycles_cost[10], pctlist[10];
+int count = 0;
+
+
+static void callback_period(struct k_timer *timer)
+{
+	u32_t pct;
+
+	stop_time = k_cycle_get_32();
+	cycles_spent = stop_time - start_time;
+	nanoseconds_spent = (u32_t)k_cyc_to_ns_floor64(cycles_spent);
+	pct = (u64_t)cycles_spent * 100000U / sys_clock_hw_cycles_per_sec();
+	start_time = stop_time;
+	pctlist[count] = pct;
+	time_spent[count] = nanoseconds_spent;
+	cycles_cost[count] = cycles_spent;
+
+	pct =0;
+
+	count++;
+	if (count >= 10) {
+		k_timer_stop(timer);
+	}
+}
+
+static void callback_duration(struct k_timer *timer)
+{
+	stop_time = k_cycle_get_32();
+}
+
+int test_periodtimer_frequency(void)
 {
-	u32_t start, end, delta, pct;
+	int i;
+
+	TC_PRINT("Testing system tick frequency\n");
+
+    k_timer_init(&test_periodtimer, callback_period, NULL);
+    start_time = k_cycle_get_32();
+    k_timer_start(&test_periodtimer, 1, 1);
+    k_sleep(1000);
+    k_timer_stop(&test_periodtimer);
+    for(i = 0; i < 10; i++)
+    {
+       printk("delta: %u expected cycles: %u "
+			   "expected time(nanosec): %u "
+			   "cost time(nanosec): %u  %u%%\n",
+			   cycles_cost[i], sys_clock_hw_cycles_per_sec()/1000,
+			   1000000 ,time_spent[i], pctlist[i]);
+
+      /* Heuristic: if we're more than 10% off, throw an error */
+       if (pctlist[i] < 90 || pctlist[i] > 110) {
+		   TC_PRINT("Clock calibration is way off!\n");
+		   return -1;
+		   break;
+	    }
+
+	     time_spent[i] = 0;
+         cycles_cost[i] = 0;
+         pctlist[i] = 0;
+      }
+
+      start_time = 0;
+      stop_time = 0;
+      cycles_spent = 0;
+      nanoseconds_spent = 0;
+
+      return 0;
+}
+
+int test_oneshottimer_frequency(void)
+{
+	u32_t delta, pct;
+
+	TC_PRINT("Testing system tick frequency\n");
+
+    k_timer_init(&test_timer, callback_duration, NULL);
+    start_time = k_cycle_get_32();
+
```

Comment:
This likely needs some kind of initial synchronization (just a `k_sleep(1);` is sufficient) to make sure you're reliably beginning the test on a tick boundary.  As is, you might start anywhere within the tick, but you're only asking for a "next tick" expiration of your timer.  It might fire in one tick, but it also might fire in just under two ticks.  Your 10% error rate doesn't handle that for all configurations.

By aligning to the tick boundary at the start of the test (assuming nothing else is going to delay your test code), you make the runtime behavior deterministic.

Reply:
I should note: in practice this may not be failing if the test before this case naturally ends on a tick boundary (as most of these timer tests do).  But that's fragile.

Reply:
Thanks for your comments, and I have updated the code to add the function:” k_sleep(1)” before the function:” k_timer_init()” is be invoked. Could you please help to review it again ? Thanks again